src/value.jl

```diff
@@ -1,145 +1,174 @@
-Base.convert(::Type{Union{CategoricalValue, Null}}, x::CategoricalValue) = x
-
-Base.promote_rule(::Type{CategoricalValue{S, R}}, ::Type{T}) where {S, T, R} = promote_type(S, T)
-Base.promote_rule(::Type{CategoricalValue{S}}, ::Type{T}) where {S, T} = promote_type(S, T)
-Base.promote_rule(::Type{CategoricalValue}, ::Type{T}) where {T} = T
+# union of all types that have "categorical value" trait
+const CatValue{R} = Union{CategoricalValue{T, R} where T,
+                          CategoricalString{R}}
+
+# checks whether the type is "categorical value"
+iscatvalue(::Type) = false
+iscatvalue(::Type{Union{}}) = false # otherwise it dispatches to Type{<:...}
+iscatvalue(::Type{<:CatValue}) = true
+iscatvalue(x::Any) = iscatvalue(typeof(x))
+
+leveltype(::Type{<:CategoricalValue{T}}) where {T} = T
+leveltype(::Type{<:CategoricalString}) = String
+leveltype(::Type) = throw(ArgumentError("Not a \"categorical value\" type"))
+leveltype(x::Any) = leveltype(typeof(x))
+# eltype() is a synonym of leveltype() for categorical values
+Base.eltype(::Type{T}) where {T <: CatValue} = leveltype(T)
+Base.eltype(x::CatValue) = eltype(typeof(x))
+
+# integer type of category reference codes used by categorical value
+reftype(::Type{<:CatValue{R}}) where {R} = R
+reftype(x::Any) = reftype(typeof(x))
+
+pool(x::CatValue) = x.pool
+level(x::CatValue) = x.level
+
+# extract the type of the original value from array eltype `T`
+unwrap_catvaluetype(::Type{T}) where {T} = T
+unwrap_catvaluetype(::Type{T}) where {T >: Null} =
+    Union{unwrap_catvaluetype(Nulls.T(T)), Null}
+unwrap_catvaluetype(::Type{Null}) = Null # to prevent dispatching to T>:Null method
+unwrap_catvaluetype(::Type{Any}) = Any # to prevent dispatching to T>:Null method
+unwrap_catvaluetype(::Type{T}) where {T <: CatValue} = leveltype(T)
+
+# get the "categorical value" type given value type `T` and reference type `R`
+catvaluetype(::Type{T}, ::Type{R}) where {T >: Null, R} =
+    catvaluetype(Nulls.T(T), R)
+catvaluetype(::Type{T}, ::Type{R}) where {T <: CatValue, R} =
+    reftype(T) === R ? T : catvaluetype(leveltype(T), R)
+catvaluetype(::Type{Any}, ::Type{R}) where {R} =
+    CategoricalValue{Any, R}  # to prevent dispatching to T>:Null method
+catvaluetype(::Type{T}, ::Type{R}) where {T, R} =
+    CategoricalValue{T, R}
+catvaluetype(::Type{<:AbstractString}, ::Type{R}) where {R} =
+    CategoricalString{R}
+# default to CategoricalString in degenerated case (all nulls)
+catvaluetype(::Type{Null}, ::Type{R}) where {R} =
+    CategoricalString{R}
+
+Base.get(x::CatValue) = index(pool(x))[level(x)]
+order(x::CatValue) = order(pool(x))[level(x)]
+
+# creates categorical value for `level` from the `pool`
+# The result is of type `C` that has "categorical value" trait
+catvalue(level::Integer, pool::CategoricalPool{T, R, C}) where {T, R, C} =
+    C(convert(R, level), pool)
+
+Base.convert(::Type{T}, x::T) where {T <: CatValue} = x
+Base.convert(::Type{Union{T, Null}}, x::T) where {T <: CatValue} = x
```

Comment:
Why is this second definition needed? Also better put these with other `convert` methods.

Reply:
I will put, but I thought those were for resolving ambiguities with Base as your comment states.

Reply:
`Base.convert(::Type{Union{T, Null}}, x::T) where {T <: CatValue} = x` is an interesting case, it's required to make `x2 == collect(x)` test to pass.
`Base.collect(CategoricalArray{Union{T,Null}})` first generates `Array{Union{Null,CategoricalValue{T}}}` and then fills it with `setindex!(::Array{Union{Null,CategoricalValue{T}}}, ::CategoricalValue{T}, ::Int)`.
So `setindex!()` calls `convert(Union{Null,CategoricalValue{T}}}, ::CategoricalValue{T})`. Normally, it should be handled by `Nulls` generic `convert`, but for categorical values there's fallback convert `Base.convert(::Type{S}, x::CatValue) where {S} = convert(S, get(x))` that seemingly overrides the Null's one as more specific (why?).
So I had to define the one in question.

I guess the fact that `collect()` returns `Array{CategoricalValue}` is not nice, and there should be `collect(CategoricalArray)` and `collect(SubArray{CategoricalArray})`.
But it should be a topic of another PR.

Reply:
Specificity issues are quite hard to get right (and there may have been some bugs)... And yes, we should probably override `collect`.